### PR TITLE
[rabit harden] fix rabit tests

### PR DESCRIPTION
**Pull request by chenqin**
**Notice, this pr will fail test unless  dmlc-core local tracker pr landed**. https://github.com/dmlc/dmlc-core/pull/510

This first pr of series of prs in order to harden rabit and make sure the tests running with meaningful test coverage. The goal of this pr is to enable model recover tests which simulate worker failure and resume (catchup with next allreduce) [expected behavior in guide doc](https://github.com/dmlc/rabit/blob/master/doc/guide.md)

What this pr does is basically working along side with local tracker pr and pass latest command into rabit init;  remove duplicated ntrail value overwrite from DMLC_NUM_ATTEMPT; add console output when worker recovered/catchup in allreduce.

### CODE REVIEW

**Comment by CodingCat** [comment]
you can consider temporarily point dmlc-core in this PR to your private branch to have a cross-check before we merge any of them

File: test/test.mk [hunk 1]
```diff
@@ -4,27 +4,27 @@
 all: model_recover_10_10k  model_recover_10_10k_die_same model_recover_10_10k_die_hard local_recover_10_10k
 
 # this experiment test recovery with actually process exit, use keepalive to keep program alive
-# TODO: enable those tests once we fix issue in rabit
 model_recover_10_10k:
-	../dmlc-core/tracker/dmlc-submit --cluster local --num-workers=10 model_recover 10000 mock=0,0,1,0 mock=1,1,1,0
+	../dmlc-core/tracker/dmlc-submit --cluster local --num-workers=2 model_recover 10000 mock=0,0,1,0 mock=1,1,1,0 rabit_num_trial=0 DMLC_NUM_ATTEMPT=10
```
[the remark this answers is not in the document]
Comment (chenqin) [reply test/test.mk]: fixed.

File: test/model_recover.cc [hunk 2]
```diff
@@ -100,26 +100,30 @@ int main(int argc, char *argv[]) {
...
     int n;
     if (sscanf(argv[i], "rabit_num_trial=%d", &n) == 1) ntrial = n;
+
   }
   int iter = rabit::LoadCheckPoint(&model);
   if (iter == 0) {
     model.InitModel(n);
-    printf("[%d] reload-trail=%d, init iter=%d\n", rank, ntrial, iter);
-  } else {
-    printf("[%d] reload-trail=%d, init iter=%d\n", rank, ntrial, iter);
```
[the remark this answers is not in the document]
Comment (chenqin) [reply test/model_recover.cc]: fixed.

File: test/model_recover.cc [hunk 3]
```diff
@@ -100,26 +100,30 @@ int main(int argc, char *argv[]) {
...
     for (int i = 0; i < nproc; i += step) {
       TestBcast(n, i, ntrial, r);
     }
     printf("[%d] !!!TestBcast pass, iter=%d\n", rank, r);
+    
+
     TestSum(&model, ntrial, r);
     printf("[%d] !!!TestSum pass, iter=%d\n", rank, r);
+
+
```
[the remark this answers is not in the document]
Comment (chenqin) [reply test/model_recover.cc]: fixed.

File: test/model_recover.cc [hunk 4]
```diff
@@ -100,26 +100,30 @@ int main(int argc, char *argv[]) {
...
     TestMax(&model, ntrial, r);
     printf("[%d] !!!TestMax pass, iter=%d\n",  rank, r);
+
     int step = std::max(nproc / 3, 1);
     for (int i = 0; i < nproc; i += step) {
       TestBcast(n, i, ntrial, r);
     }
     printf("[%d] !!!TestBcast pass, iter=%d\n", rank, r);
+    
+
```
[the remark this answers is not in the document]
Comment (chenqin) [reply test/model_recover.cc]: fixed.

**Comment by chenqin** [comment]
passing test locally with gcc-7/g++-7 on osx 10.14.3.

**Comment by chenqin** [comment]
^^ @CodingCat @hcho3

**Comment by chenqin** [comment]
Passed integration tests in xgboost side
https://travis-ci.org/chenqin/xgboost/builds/505366624?utm_medium=notification&utm_source=email

File: src/allreduce_robust.cc [hunk 5]
```diff
@@ -94,6 +94,11 @@ void AllreduceRobust::Allreduce(void *sendrecvbuf_,
     return;
   }
   bool recovered = RecoverExec(sendrecvbuf_, type_nbytes * count, 0, seq_counter);
+
+  // check if restarted worker recovered from other healthy workers
+  if (recovered) {
+      printf("[%d] all reduce with recover with seq %d\n", this->rank, seq_counter);
```
Comment (CodingCat) [inline src/allreduce_robust.cc]: hard to read here....[$this->rank]  recovered with seq_id $seq_counter?

File: src/allreduce_robust.cc [hunk 6]
```diff
@@ -270,6 +275,9 @@ void AllreduceRobust::CheckPoint_(const Serializable *global_model,
       if (RecoverExec(NULL, 0, 0, ActionSummary::kLocalCheckPoint)) break;
       // save model model to new version place
```
Comment (CodingCat) [inline src/allreduce_robust.cc]: some typo here?

File: src/allreduce_robust.cc [hunk 7]
```diff
@@ -270,6 +275,9 @@ void AllreduceRobust::CheckPoint_(const Serializable *global_model,
...
       int new_version = !local_chkpt_version;
+      printf("[%d] local checkpoint # %d, new # %d \n",
+              this->rank, local_chkpt_version, new_version);
```
Comment (CodingCat) [inline src/allreduce_robust.cc]: I think this line does covey anything beyond what Line 307 says,,,we may only need one of them

File: src/allreduce_robust.cc [hunk 8]
```diff
@@ -737,6 +746,9 @@ AllreduceRobust::ReturnType AllreduceRobust::TryLoadCheckPoint(bool requester) {
     succ = TryRecoverLocalState(&local_rptr[local_chkpt_version],
                                 &local_chkpt[local_chkpt_version]);
     if (succ != kSuccess) return succ;
+
+    printf("[%d] recovered local checkpoint version %d \n", this->rank, local_chkpt_version);
```
Comment (CodingCat) [inline src/allreduce_robust.cc]: recovered from?

File: Makefile [hunk 9]
```diff
@@ -2,7 +2,7 @@ OS := $(shell uname)
 
 ifeq ($(OS), Linux)
 ifndef CXX
-export CXX = g++
+export CXX = g++-5
```
[the remark this answers is not in the document]
Comment (chenqin) [reply Makefile]: reverted, now use gcc g++ 4.8

File: src/allreduce_robust.cc (hunk 5 above)
Comment (chenqin) [reply src/allreduce_robust.cc]: removed, will rethink how to log debug recovery in separate pr

File: src/allreduce_robust.cc (hunk 6 above)
Comment (chenqin) [reply src/allreduce_robust.cc]: fixed

File: src/allreduce_robust.cc (hunk 7 above)
Comment (chenqin) [reply src/allreduce_robust.cc]: removed

File: src/allreduce_robust.cc (hunk 8 above)
Comment (chenqin) [reply src/allreduce_robust.cc]: fixed.

**Comment by chenqin** [comment]
passing XGBoost tests 
https://travis-ci.org/chenqin/xgboost/builds/505580249

**Comment by chenqin** [comment]
@trivialfis Can you help merge this? should be straight forward pr.

**Comment by trivialfis** [comment]
@chenqin No problem.  Just one small question, is this compiling on windows?  Last time I tried to upgrade rabit in XGBoost the compilation failed  #80.

**Comment by chenqin** [comment]
should be good
https://ci.appveyor.com/project/chenqin/xgboost/builds/23031675

**Pull request merged**